apollo-router/src/spec/field_type.rs

```diff
@@ -80,114 +82,83 @@ impl std::fmt::Display for FieldType {
-                Ok(())
-            } else {
-                Err(InvalidValue)
-            }
-        }
-        (hir::Type::Named { name, .. }, Value::Bool(_)) if name == "Boolean" => Ok(()),
-        (hir::Type::List { ty: inner_ty, .. }, Value::Array(vec)) => vec
-            .iter()
-            .try_for_each(|x| validate_input_value(inner_ty, x, schema)),
-        // For coercion from single value to list
-        (hir::Type::List { ty: inner_ty, .. }, val) if val != &Value::Null => {
-            validate_input_value(inner_ty, val, schema)
-        }
-        (hir::Type::NonNull { ty: inner_ty, .. }, value) => {
-            if value.is_null() {
-                Err(InvalidValue)
-            } else {
-                validate_input_value(inner_ty, value, schema)
-            }
-        }
-        (hir::Type::Named { name, .. }, _)
-            if schema
-                .type_system
-                .definitions
-                .scalars
-                .get(name)
-                .map(|def| !def.is_built_in())
-                .unwrap_or(false)
-                || schema.type_system.definitions.enums.contains_key(name) =>
-        {
-            Ok(())
-        }
-        // NOTE: graphql's types are all optional by default
-        (_, Value::Null) => Ok(()),
-        (hir::Type::Named { name, .. }, value) => {
-            if let Some(value) = value.as_object() {
-                if let Some(object_ty) = schema.type_system.definitions.input_objects.get(name) {
-                    object_ty.fields().try_for_each(|field| {
-                        let default: Value;
-                        let value = match value.get(field.name()) {
-                            Some(&Value::Null) | None => {
-                                default = field
-                                    .default_value()
-                                    .and_then(parse_hir_value)
-                                    .unwrap_or(Value::Null);
-                                &default
-                            }
-                            Some(value) => value,
-                        };
-                        validate_input_value(field.ty(), value, schema)
-                    })
-                } else {
-                    Err(InvalidValue)
-                }
-            } else {
-                Err(InvalidValue)
-            }
+        "ID" => return from_bool(value.is_valid_id_input()),
+        "Boolean" => return from_bool(value.is_boolean()),
+        _ => {}
+    }
+    let type_def = schema
+        .definitions
+        .types
+        .get(type_name)
+        .ok_or(InvalidValue)?;
+    match (type_def, value) {
+        // Custom scalar: accept any JSON value
+        (schema::ExtendedType::Scalar(_), _) => Ok(()),
+
+        // TODO: check enum value?
+        // (schema::ExtendedType::Enum(def), Value::String(s)) => {
+        //     from_bool(def.values.contains_key(s))
+        // },
```

Comment:
do we have followup tickets for todos in this file ?

Reply:
https://github.com/apollographql/router/issues/4040